File: frontend/src/mutationEngine.js
const SNAPSHOT_HEADER_PREFIXES = ["classdiagram", "flowchart", "graph"];

const STYLE_LINE_PATTERN =
  /^(style\s|classdef\s|class\s+\w+\s+\w+|linkstyle\s|click\s)/i;

export function normalizeCanvas(text) {
  const stripped = text?.trim() ?? "";
  if (!stripped) return "";
  return stripped.endsWith("\n") ? stripped : `${stripped}\n`;
}

export function isSnapshotMutation(mutation) {
  const lines = mutation
    .trim()
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean);
  if (!lines.length) return false;
  const first = lines[0].toLowerCase();
  return SNAPSHOT_HEADER_PREFIXES.some((prefix) => first.startsWith(prefix));
}

export function isStyleMutation(mutation) {
  const lines = mutation
    .trim()
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean);
  if (!lines.length) return false;
  return lines.every((line) => STYLE_LINE_PATTERN.test(line));
}

export function appendLines(canvas, mutation) {
  let result = canvas;
  for (const line of mutation.trim().split("\n")) {
    const stripped = line.trim();
    if (!stripped || result.includes(stripped)) continue;
    if (result && !result.endsWith("\n")) result += "\n";
    result += `${stripped}\n`;
  }
  return result;
}

export function applyMutationMode(canvas, mutation) {
  if (!mutation?.trim()) {
    return { canvas, mode: "incremental" };
  }

  if (isSnapshotMutation(mutation)) {
    return { canvas: normalizeCanvas(mutation), mode: "snapshot" };
  }

  if (isStyleMutation(mutation)) {
    return { canvas: appendLines(canvas, mutation), mode: "style" };
  }

  return { canvas, mode: "incremental" };
}

export function resolveLevelCanvas(config, levelIndex) {
  const level = config.levels?.[levelIndex - 1];
  if (level?.initial_canvas) return level.initial_canvas;
  return config.initial_canvas;
}

File: backend/mutation_engine.py
"""Unified canvas mutation: snapshot overwrite, style append, or incremental merge."""

from __future__ import annotations

import re

SNAPSHOT_HEADER_PREFIXES = (
    "classdiagram",
    "flowchart",
    "graph",
)

STYLE_LINE_PATTERN = re.compile(
    r"^(style\s|classdef\s|class\s+\w+\s+\w+|linkstyle\s|click\s)",
    re.IGNORECASE,
)


def normalize_canvas(text: str) -> str:
    stripped = text.strip()
    if not stripped:
        return ""
    return stripped + ("\n" if not stripped.endswith("\n") else "")


def is_snapshot_mutation(mutation: str) -> bool:
    """Solution 1: mutation is a full diagram replacement."""
    lines = [line.strip() for line in mutation.strip().splitlines() if line.strip()]
    if not lines:
        return False
    first = lines[0].lower()
    return any(first.startswith(prefix) for prefix in SNAPSHOT_HEADER_PREFIXES)


def is_style_mutation(mutation: str) -> bool:
    """Solution 3: append-only style/class overrides for existing nodes."""
    lines = [line.strip() for line in mutation.strip().splitlines() if line.strip()]
    if not lines:
        return False
    return all(STYLE_LINE_PATTERN.match(line) for line in lines)


def append_lines(canvas: str, mutation: str) -> str:
    result = canvas
    for line in mutation.strip().splitlines():
        stripped = line.strip()
        if not stripped or stripped in result:
            continue
        if result and not result.endswith("\n"):
            result += "\n"
        result += stripped + "\n"
    return result


def apply_mutation_mode(canvas: str, mutation: str) -> tuple[str, str]:
    """
    Apply mutation and return (new_canvas, mode).
    mode is one of: snapshot, style, incremental.
    """
    if not mutation or not mutation.strip():
        return canvas, "incremental"

    if is_snapshot_mutation(mutation):
        return normalize_canvas(mutation), "snapshot"

    if is_style_mutation(mutation):
        return append_lines(canvas, mutation), "style"

    return canvas, "incremental"
